### ISSUE TITLE
grouped_gemm_fp8_rowwise inconsistent overwriting of output data.

### ISSUE THREAD

**Issue by HDCharles** [pytorch/FBGEMM#4314]
i've been benchmarking and testing the grouped_gemm_fp8_rowwise kernel compared to the pytorch core grouped_scaled_mm but I ran into issues with accuracy. My results were extremely inconsistent but after some digging i found that subsequent invocations of grouped_gemm_fp8_rowwise seem to overwrite the old output even if i clone or detach the original output.

The situation where this comes up is in the main calculation of a moe, you normally do something like

```
y1=grouped_gemm(....).clone()
y3 = grouped_gemm(....).clone() # <- the calculation of y3 changes the value of y1.
y=silu(y1)*y3

...
```

when i used the fbgemm kernel for the grouped gemm, the y3 calculation overwrites the value of the y1 calculation making y have a value of 0. Note it tried to make a simple repro but the behavior seemed to be kind of finicky to reproduce. I had to use some specific setup to get it to repro, if you just use random fp8 tensors, i couldn't get it to exibit the behavior.

I tried this on cuda 12.6 on an H100 with both the nightly and stable fbgemm versions and both showed the same behavior

```

import pytest
import torch

from torchao.prototype.moe_quant.quantizable_moe_modules import (
    MOEFeedForwardAOQuantizable,
)
from torchao.prototype.moe_quant.utils import (
    MoEQuantConfig,
    cond_ffn_filter,
)
from torchao.quantization.quant_api import (
    Float8DynamicActivationFloat8WeightConfig,
    PerRow,
    quantize_,
)

if torch.version.hip is not None:
    pytest.skip(
        "ROCm support for MoE quantization is under development",
        allow_module_level=True,
    )
import torch.nn.functional as F
from fbgemm_gpu.experimental.gemm.triton_gemm.grouped_gemm import (
    grouped_gemm_fp8_rowwise,
)

torch.manual_seed(0)


def manual_pad(num_tokens_per_expert, alignment):
    num_experts = num_tokens_per_expert.shape[0]

    padded_sizes = (
        (num_tokens_per_expert + (num_tokens_per_expert == 0)) / alignment
    ).ceil() * alignment
    pad_len = int(padded_sizes.sum().item())

    padded_indices = (
        torch.zeros(pad_len, dtype=torch.int32, device=num_tokens_per_expert.device) - 1
    )
    start_tok_index = 0
    start_pad_index = 0
    for i in range(num_experts):
        end_tok_index = int(start_tok_index + num_tokens_per_expert[i].item())
        end_pad_index = int(start_pad_index + num_tokens_per_expert[i].item())
        padded_indices[start_pad_index:end_pad_index] = torch.arange(
            start_tok_index,
            end_tok_index,
            dtype=torch.int32,
            device=num_tokens_per_expert.device,
        )
        start_tok_index = end_tok_index
        start_pad_index = start_pad_index + int(padded_sizes[i].item())
    m_offsets = padded_sizes.cumsum(0).to(torch.int32)
    return padded_indices, m_offsets


num_tokens = 8
DEFAULT_PARAMS = (512, 256, 8, 2)

device = "cuda"
dtype = torch.bfloat16

config = MoEQuantConfig(Float8DynamicActivationFloat8WeightConfig(granularity=PerRow()))

model_params = DEFAULT_PARAMS

input_shape = (num_tokens, model_params[0])
input = torch.randn(input_shape, dtype=torch.bfloat16, device=device)

model = MOEFeedForwardAOQuantizable(*model_params, empty_init=False)
model = model.to(dtype).to(device)

out_orig = model(input)

quantize_(model, config, cond_ffn_filter)

w1 = model.experts.w1
w2 = model.experts.w2
w3 = model.experts.w3

router = model.router
top_k = model.top_k

# preprocess
scores = router(input)  # [T, E]
scores = torch.nn.functional.softmax(scores, dim=-1)
scores, expert_indices = torch.topk(scores, top_k, dim=-1)  # [T, A], [T, A]
scores /= scores.sum(dim=-1, keepdim=True).to(input.dtype)  # [T, A]

use_fbgemm_kernel = True
fast_accum = True

orig_in_shape = input.shape
input.reshape(-1, orig_in_shape[-1])
num_tokens, dim = input.shape
num_experts, expert_dim, _ = w1.shape
scores = scores.view(-1, scores.shape[-1])
top_k = scores.shape[-1]
total_activations = num_tokens * top_k

# preprocess indices
expert_indices = expert_indices.view(-1)
activation_shuffle = expert_indices.argsort(stable=True)
token_shuffle = activation_shuffle.div(top_k).floor().to(torch.int64)
num_tokens_per_expert = torch.histc(
    expert_indices, bins=num_experts, min=0, max=num_experts
)

# padding
alignment = 16
padded_indices, m_offsets = manual_pad(num_tokens_per_expert, alignment)
pad_len = padded_indices.shape[0]
valid_values = padded_indices >= 0

# get data for weights
w1_fp8 = w1.original_weight_tensor.tensor_impl.float8_data
w1_scale = w1.original_weight_tensor.tensor_impl.scale.squeeze()
w1_qfunc = w1.input_quant_func
w1_quant_kwargs = w1.quant_kwargs

w3_fp8 = w3.original_weight_tensor.tensor_impl.float8_data
w3_scale = w3.original_weight_tensor.tensor_impl.scale.squeeze()

w2_fp8 = w2.original_weight_tensor.tensor_impl.float8_data
w2_scale = w2.original_weight_tensor.tensor_impl.scale.squeeze()
w2_qfunc = w2.input_quant_func
w2_quant_kwargs = w2.quant_kwargs


# quantize then shuffle input
q_input = w1_qfunc(input, **w1_quant_kwargs)
q_input_data = q_input.tensor_impl.float8_data
q_input_scale = q_input.tensor_impl.scale.squeeze()
input_fp8 = torch.zeros(
    (pad_len, q_input_data.shape[-1]),
    dtype=q_input_data.dtype,
    device=q_input_data.device,
)
input_scale = torch.zeros(
    pad_len, dtype=q_input_scale.dtype, device=q_input_scale.device
)
input_fp8[valid_values] = q_input_data[token_shuffle]
input_scale[valid_values] = (
    q_input_scale[token_shuffle] if q_input_scale.numel() > 1 else q_input_scale
)

if use_fbgemm_kernel:
    assert grouped_gemm_fp8_rowwise is not None, (
        "fbgemm kernel requires fbgemm-gpu-genai to be installed: https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/experimental/gen_ai/README.md"
    )

    for i in range(10):
        y1 = (
            grouped_gemm_fp8_rowwise(
                input_fp8,
                w1_fp8.reshape(-1, w1_fp8.shape[-1]),
                m_offsets,
                input_scale,
                w1_scale.reshape(-1),
                use_fast_accum=True,
            )
            .clone()
            .detach()
            + 0
        )
        print("a", y1.sum())
        y3 = (
            grouped_gemm_fp8_rowwise(
                input_fp8,
                w3_fp8.reshape(-1, w3_fp8.shape[-1]),
                m_offsets,
                input_scale,
                w3_scale.reshape(-1),
                use_fast_accum=True,
            )
            .clone()
            .detach()
            + 0
        )
        torch.cuda.synchronize()
        y = F.silu(y1) * y3
        print("b", y1.sum(), y3.sum())
        y = F.silu(y1) * y3

```

output: 

a tensor(784., device='cuda:0', dtype=torch.bfloat16)

b tensor(784., device='cuda:0', dtype=torch.bfloat16) tensor(-1368., device='cuda:0', dtype=torch.bfloat16)

---- ^ first iteration works fine

a tensor(784., device='cuda:0', dtype=torch.bfloat16)  ########while hte initial output had a value of 784...

b tensor(0., device='cuda:0', dtype=torch.bfloat16) tensor(-1368., device='cuda:0', dtype=torch.bfloat16)

----- ^ second iteration doens't work though, once the other calculation is performed the value of the first output becomes 0

a tensor(784., device='cuda:0', dtype=torch.bfloat16)

b tensor(784., device='cuda:0', dtype=torch.bfloat16) tensor(-1368., device='cuda:0', dtype=torch.bfloat16)

a tensor(784., device='cuda:0', dtype=torch.bfloat16)

env info:

```
❯ python -m torch.utils.collect_env Output
<frozen runpy>:128: RuntimeWarning: 'torch.utils.collect_env' found in sys.modules after import of package 'torch.utils', but prior to execution of 'torch.utils.collect_env'; this may result in unpredictable behaviour
Collecting environment information...
PyTorch version: 2.8.0.dev20250603+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5)
Clang version: 20.1.1 (CentOS 20.1.1-3.el9)
CMake version: version 3.31.6
Libc version: glibc-2.34

Python version: 3.12.9 | packaged by Anaconda, Inc. | (main, Feb  6 2025, 18:56:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.4.3-0_fbk20_zion_2830_g3e5ab162667d-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.6.85
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100
GPU 1: NVIDIA H100
GPU 2: NVIDIA H100
GPU 3: NVIDIA H100
GPU 4: NVIDIA H100
GPU 5: NVIDIA H100
GPU 6: NVIDIA H100
GPU 7: NVIDIA H100

Nvidia driver version: 535.154.05
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.9.6.0
/usr/lib64/libcudnn_adv.so.9.6.0
/usr/lib64/libcudnn_cnn.so.9.6.0
/usr/lib64/libcudnn_engines_precompiled.so.9.6.0
/usr/lib64/libcudnn_engines_runtime_compiled.so.9.6.0
/usr/lib64/libcudnn_graph.so.9.6.0
/usr/lib64/libcudnn_heuristic.so.9.6.0
/usr/lib64/libcudnn_ops.so.9.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             384
On-line CPU(s) list:                0-383
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 9654 96-Core Processor
CPU family:                         25
Model:                              17
Thread(s) per core:                 2
Core(s) per socket:                 96
Socket(s):                          2
Stepping:                           1
Frequency boost:                    enabled
CPU(s) scaling MHz:                 78%
CPU max MHz:                        3707.8120
CPU min MHz:                        1500.0000
BogoMIPS:                           4793.19
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d
Virtualization:                     AMD-V
L1d cache:                          6 MiB (192 instances)
L1i cache:                          6 MiB (192 instances)
L2 cache:                           192 MiB (192 instances)
L3 cache:                           768 MiB (24 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-95,192-287
NUMA node1 CPU(s):                  96-191,288-383
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Vulnerable: eIBRS with unprivileged eBPF
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] mypy==1.14.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.2
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnx==1.17.0
[pip3] onnxscript==0.2.2
[pip3] optree==0.13.0
[pip3] pytorch-triton==3.3.1+gitc8757738
[pip3] torch==2.8.0.dev20250603+cu126
[pip3] torchao==0.11.0+git22e19a43
[pip3] torchaudio==2.6.0.dev20250530+cu126
[pip3] torchdata==0.11.0
[pip3] torchtitan==0.0.2
[pip3] torchtune==0.0.0
[pip3] torchvision==0.22.0.dev20250530+cu126
[pip3] triton==3.2.0
[conda] magma-cuda126             2.6.1                         1    pytorch
[conda] mkl-include               2025.0.1                 pypi_0    pypi
[conda] mkl-static                2025.0.1                 pypi_0    pypi
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.6.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.6.80                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.6.77                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.6.77                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.5.1.17                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.0.4                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.7.77                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.1.2                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.4.2                 pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.3                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.26.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.85                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.6.77                  pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] pytorch-triton            3.3.1+gitc8757738          pypi_0    pypi
[conda] torch                     2.8.0.dev20250603+cu126          pypi_0    pypi
[conda] torchao                   0.11.0+git22e19a43           dev_0    <develop>
[conda] torchaudio                2.6.0.dev20250530+cu126          pypi_0    pypi
[conda] torchdata                 0.11.0                   pypi_0    pypi
[conda] torchtitan                0.0.2                    pypi_0    pypi
[conda] torchtune                 0.0.0                    pypi_0    pypi
[conda] torchvision               0.22.0.dev20250530+cu126          pypi_0    pypi
[conda] triton                    3.2.0                    pypi_0    pypi
```

**Comment by moderato** [comment]
Hey there I'm also debugging the `grouped_gemm_fp8_rowwise` function but I have a different error (segfault when `_use_warp_specialization` is set to True). I saw you are using triton 3.2.0. Is that ws-3.2.x as suggested in the code (https://github.com/pytorch/FBGEMM/blob/6768baf605a3434aeabb2f2e43c9caf8d3eecc19/fbgemm_gpu/experimental/gemm/triton_gemm/grouped_gemm.py#L976)?

**Comment by HDCharles** [comment]
i think you may need to use the nightly or something, i'm not trying to use the warp specialization

**Comment by HDCharles** [comment]
#resolved I had a mistake in my code above the input to the function is sizes rather than offsets. Its a bit worrying that the kernel actually ran when the size info was so mangled but changing it at least makes it work as expected
